### PR TITLE
Optimizers

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
Added RMSprop and Adagrad optimizers and corresponding unit tests.

### CODE REVIEW

**Comment by dsyme** [comment]
Looks great!

File: src/Native/LibTorchSharp/THSNN.cpp [hunk 1]
```diff
@@ -499,19 +499,34 @@ Tensor THSNN_poisson_loss(const Tensor input, const Tensor target, const bool lo
...
+{
+    auto params = toTensors<at::Tensor>((torch::Tensor**)parameters, length);
+    auto options = torch::optim::AdagradOptions(learning_rate).lr_decay(lr_decay).weight_decay(weight_decay);
+
+    return new std::shared_ptr<torch::optim::Optimizer>(std::make_shared<torch::optim::Adagrad>(torch::optim::Adagrad(params, options)));
+}
+
+Optimizer THSNN_Adam_ctor(const Tensor* parameters, const int length, const double learning_rate)
+{
+    auto  params = toTensors<at::Tensor>((torch::Tensor**)parameters, length);
```
Comment (dsyme) [inline src/Native/LibTorchSharp/THSNN.cpp:512]: There is an AdamOptions we should be using here https://pytorch.org/cppdocs/api/structtorch_1_1optim_1_1_adam_options.html?highlight=adamoptions#_CPPv4N5torch5optim11AdamOptionsE

Comment (dsyme) [reply src/Native/LibTorchSharp/THSNN.cpp:512]: Also this has numerous additional parameters available on it

Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSNN.cpp:512]: Yes, we can add that. There are a lot of options that are being skipped for the various optimizers. I can add them systematically.

File: src/Native/LibTorchSharp/THSNN.cpp [hunk 2]
```diff
@@ -499,19 +499,34 @@ Tensor THSNN_poisson_loss(const Tensor input, const Tensor target, const bool lo
     )
 }
 
-Optimizer THSNN_Adam_ctor(const Tensor* parameters, const int length, const double learnig_rate)
+Optimizer THSNN_Adagrad_ctor(const Tensor* parameters, const int length, const double learning_rate, const double lr_decay, const double weight_decay)
+{
+    auto params = toTensors<at::Tensor>((torch::Tensor**)parameters, length);
+    auto options = torch::optim::AdagradOptions(learning_rate).lr_decay(lr_decay).weight_decay(weight_decay);
```
Comment (dsyme) [inline src/Native/LibTorchSharp/THSNN.cpp:505]: What happens about the optionality of these?  

We don't have any consistent way of passing across optional values across the C API, just adhoc uses of NULL, min values and so on

File: src/TorchSharp/NN/Optimizer.cs [hunk 3]
```diff
@@ -78,10 +78,40 @@ protected void Dispose (bool disposing)
...
+        /// <summary>
+        /// Implements RMSprop algorithm.
+        ///
+        /// Proposed by G.Hinton in his course.
+        /// </summary>
+        /// <param name="parameters">Prameters to optimize</param>
+        /// <param name="learningRate">Learning rate (default: 1e-2)</param>
+        /// <param name="alpha">Smoothing constant (default: 0.99)</param>
+        /// <returns></returns>
+        public static Optimizer RMSProp(IEnumerable<TorchTensor> parameters, double learningRate = 0.01, double alpha = 0.99)
```
Comment (dsyme) [inline src/TorchSharp/NN/Optimizer.cs:93]: I assume you double checked the default valus with the pytorch docs?

File: src/Native/LibTorchSharp/THSNN.cpp (hunk 2 above)
Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSNN.cpp:505]: We should remove the defaults when crossing the interop boundary.

File: src/TorchSharp/NN/Optimizer.cs (hunk 3 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Optimizer.cs:93]: Yes.

File: src/Native/LibTorchSharp/THSNN.cpp (hunk 1 above)
Comment (dsyme) [reply src/Native/LibTorchSharp/THSNN.cpp:512]: Thanks.  If you skip anything knowingly please make sure there's a note of some form in an open bug e.g. just add it to one of the existing bugs listing missing things

File: src/Native/LibTorchSharp/THSNN.cpp (hunk 2 above)
Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSNN.cpp:505]: I'll do that next.

Comment (dsyme) [reply src/Native/LibTorchSharp/THSNN.cpp:505]: OK yes in this case that can be done.  

In some other cases I've so far made the optional explicit, e.g. see 
* `THSTensor_avg_pool2d_backward` `divisor_override` 
* `THSTensor_fft` 
* `THSTensor_upsample_nearest1d` - `outputSize` and `scaleFactors`

and so on. There are about 25 explicit uses of `c10::optional`

If you think it's better to simplify these out to optionality on the C# side please go ahead

**Pull request merged**